src/validation.cpp

```diff
@@ -4903,33 +4933,78 @@ static CDiskBlockPos SaveBlockToDisk(const CBlock& block, int nHeight, const CCh
     return blockPos;
 }
 
-bool CChainState::ContextualCheckZerocoinStake(CBlockIndex* pindex, CStakeInput* stake)
+bool CChainState::ContextualCheckStake(CBlockIndex* pindex, CStakeInput* stake)
 {
-    if (ZerocoinStake* stakeCheck = dynamic_cast<ZerocoinStake*>(stake)) {
-        CBlockIndex* pindexFrom = stakeCheck->GetIndexFrom();
-        if (!pindexFrom)
-            return error("%s: failed to get index associated with zerocoin stake checksum", __func__);
-
-        int nRequiredDepth = Params().Zerocoin_RequiredStakeDepth();
-        if (pindex->nHeight >= Params().HeightLightZerocoin())
-            nRequiredDepth = Params().Zerocoin_RequiredStakeDepthV2();
-
-        if (pindex->nHeight - pindexFrom->nHeight < nRequiredDepth)
-            return error("%s: zerocoin stake does not have required confirmation depth", __func__);
-
-        //The checksum needs to be the exact checksum from the modifier height
-        libzerocoin::CoinDenomination denom = libzerocoin::AmountToZerocoinDenomination(stakeCheck->GetValue());
-        int nHeightStake = pindex->nHeight - nRequiredDepth;
-        CBlockIndex* pindexFrom2 = pindex->GetAncestor(nHeightStake);
-        if (!pindexFrom2)
-            return error("%s: block ancestor does not exist", __func__);
-
-        uint256 hashCheckpoint = pindexFrom2->GetAccumulatorHash(denom);
-        if (hashCheckpoint != stakeCheck->GetChecksum())
-            return error("%s: accumulator checksum is different than the modifier block. indexfromheight=%d stake=%s blockfrom=%s", __func__, pindexFrom->nHeight, stakeCheck->GetChecksum().GetHex(), hashCheckpoint.GetHex());
-    } else {
-        return error("%s: dynamic_cast of stake ptr failed", __func__);
+    switch (stake->GetType()) {
+        case STAKE_ZEROCOIN:
+            ZerocoinStake* stakeZerocoin;
+            try {
+                stakeZerocoin = dynamic_cast<ZerocoinStake*>(stake);
+            } catch (std::bad_cast) {
+                return false;
+            }
```

Comment:
`dynamic_cast` on a pointer does not throw `std::bad_cast` on failure but returns a null pointer instead.